### DISCUSSION TITLE
프로젝트 - 이정수

### DISCUSSION THREAD

**Discussion by sw930718** [Pseudo-Lab/EfficientML#17] [Show and tell]
이미지 분류 Task에 대해서 다양한 기법들의 성능을 실험해 보고자 한다. 
[ 대분류 ]
Pruning / Quantization / Neural Architecture Search / Knowledge Distillation

**Comment by sw930718** [comment]
## 공통

시점 `Pre-training` `During-traing` `Post-training` 

파이프라인 `One-shot` `Iterative` 

스케쥴링 `AGP`

## Pruning 기법

![pruning 기법](https://github.com/user-attachments/assets/360c6e20-4857-48b4-9d38-6b4b039ac2cf)
_(Notion에 작성한 내용. 추후 표로 정리해 재업로드 예정)_

**Comment by sw930718** [comment]
### [ 실험1 ] 25.05.20

`import torch.nn.utils.prune as prune` & `import torch_pruning as tp`
- 중요도 낮은 가중치 제거하는 방식으로 pruning 진행

[Original] Params: 25,557,032
[Pruned] Params: 6,401,040
Reduction: 75.0%

![image (3)](https://github.com/user-attachments/assets/28a2755b-7897-45d6-9d39-a59da4bc78fe)
(전)sample 0: [('laptop, laptop computer', '11.8%'), ('desk', '10.5%'), ('notebook, notebook computer', '6.9%'), ('screen, CRT screen', '2.9%'), ('mouse, computer mouse', '2.6%')]
(후)sample 0: [('Labrador retriever', '2.2%'), ('bolo tie, bolo, bola tie, bola', '2.1%'), ('Angora, Angora rabbit', '2.1%'), ('otter', '1.1%'), ('kelpie', '1.1%')]
          
![image (4)](https://github.com/user-attachments/assets/ca6816d9-64f7-4b6b-aae3-5883a0b591b0)
(전)sample 1: [('mashed potato', '25.3%'), ('plate', '10.3%'), ('meat loaf, meatloaf', '8.6%'), ('chocolate sauce, chocolate syrup', '6.7%'), ('broccoli', '6.2%')]
(후)sample 1: [('bolo tie, bolo, bola tie, bola', '2.3%'), ('Angora, Angora rabbit', '2.0%'), ('Labrador retriever', '1.5%'), ('kelpie', '1.4%'), ('otter', '1.3%')]

### 회고
Pruning 비율을 너무 높게 잡은 탓인지, 분류 성능이 매우 떨어졌다. 
비율 조정 / 다른 기법과의 조합 등을 통해 추가 실험 예정

**Comment by namgyu-youn** [comment]
내용 잘 봤습니다. `torch.nn.utils.prune`, `torch_pruning`은 모두 Ln-norm 기반으로 rebudant weight를 측정하는 방식이라고 알고 있습니다. 즉, magnitude 만을 사용해 rebudant weight하는 방식이라고 볼 수 있을 것 같습니다. 따라서 이러한 한계를 극복하기 위해서 gradient 기반의 다양한 방법론이 있는 것 같습니다. ([[1]](https://github.com/issues/recent?issue=pytorch%7Cpytorch%7C157218))

PyTorch 경우에는 pruning과 관련된 기본적인 (magnitude-based) 기능들만 제공하고, 보다 심층적인 알고리즘은 [torchao](https://docs.pytorch.org/ao/stable/sparsity.html) 에서 이루어지는 것 같습니다. 혹시 관심이 있으시다면, `torchao`를 사용해보시는 것을 추천드리고 싶습니다.
